Time shift difference

betodealmeida opened the pull request:
I added a new control to allow comparing comparing the main time series with time shifts. In addition to showing the time shift as individual lines, it's now possible to compute absolute and relative differences.

Current behavior, will be the default option in the selector ("individual lines"):

<img width="1680" alt="screen shot 2018-06-11 at 4 03 21 pm" src="https://user-images.githubusercontent.com/1534870/41261806-4c49bf02-6d92-11e8-80cd-2aa4d51e2b4b.png">

Absolute difference:

<img width="1680" alt="screen shot 2018-06-11 at 4 03 28 pm" src="https://user-images.githubusercontent.com/1534870/41261805-4c30cda8-6d92-11e8-9fbd-9325b2845e1d.png">

Relative difference, computed as `(main - time_shift) / time_shift` (for a "1 week" shift" this means that `100` means the value today is double the value last week, and `-50` means that the value today is half of what it was last week):

<img width="1680" alt="screen shot 2018-06-11 at 4 03 37 pm" src="https://user-images.githubusercontent.com/1534870/41261804-4c1758f0-6d92-11e8-8c01-4bd91afe7ad2.png">

betodealmeida committed:
Allow different comparisons when using time shift

betodealmeida committed:
Remove test code

betodealmeida committed:
Remove lead/trail NaN and improve code

betodealmeida commented:
> I think this is effectively the same as a superior Period Ratio isn't it? If so we should db migrate to translate the slices.

I'm not familiar with the period ratio, and it doesn't seem to map 1:1 to the comparison types. Should I create a migration script only for those cases that can be mapped?

> I know one difference in the backend is that Period Ratio runs 2 queries and merges them. So I could look at the past 28 days and compare with last year's same period. I'm not sure if Time Shift can do that or it would query/show more than 28 days (which may be ok). What do you think?

How I would that with the period ratio? For time shift you would select the last 28 days and then use "1 year" as the time shift.

vylc commented:
Some updates to the AA section to improve usability
<img width="953" alt="screen shot 2018-06-12 at 3 33 09 pm" src="https://user-images.githubusercontent.com/33429480/41320960-a81df5ea-6e56-11e8-8fdc-9b4b62c1bbe6.png">

betodealmeida committed:
Add headers/subheader

betodealmeida committed:
Rebase

betodealmeida committed:
Merge branch 'master' of github.com:apache/incubator-superset into time_shift_difference

mistercrunch commented on superset/assets/src/explore/controls.jsx:
```diff
@@ -1562,24 +1562,6 @@ export const controls = {
     description: t('Compute the contribution to the total'),
   },
 
-  num_period_compare: {
```
If deprecating `num_period_compare` and `period_ratio_type`, do we need a db migration for existing charts that have been saved using that feature?

mistercrunch commented:
LGTM apart from the db migration needed comment.

betodealmeida committed:
Merge branch 'master' of github.com:apache/incubator-superset into time_shift_difference

betodealmeida committed:
Update yarn

betodealmeida committed:
Migration script

betodealmeida committed:
Fix migration

betodealmeida committed:
Small fixes

betodealmeida commented:
@mistercrunch, I've added a migration script. Here's a dashboard that was saved before the migration:

<img width="1680" alt="screen shot 2018-07-06 at 6 45 00 pm" src="https://user-images.githubusercontent.com/1534870/42405658-a9d057ce-814e-11e8-826b-676c238048ba.png">

Note that:

- granularity is `month`;
- `num_period_compare` is 24 (months);
- `period_ratio_type` is "growth" (ie, `new/old - 1`).

After the migration:

<img width="1680" alt="screen shot 2018-07-06 at 8 04 02 pm" src="https://user-images.githubusercontent.com/1534870/42406214-c28016e8-8157-11e8-84a9-024953817714.png">

Now:

- `time_compare` is "2 years" (though it doesn't show in the dropdown since it's free form);
- `comparison_type` is "percentage" (same as "growth", just renamed).

betodealmeida committed:
Merge branch 'master' of github.com:apache/incubator-superset into time_shift_difference

betodealmeida committed:
Trigger tests

betodealmeida committed:
Merge

betodealmeida committed:
Fix lint

mistercrunch commented:
LGTM

betodealmeida committed:
Delete package-lock.json

betodealmeida committed:
Fix javascript

williaster commented:
@betodealmeida this broke my local build, the db migration fails

williaster commented on superset/assets/package.json:
```diff
@@ -68,6 +68,7 @@
     "geojson-extent": "^0.3.2",
     "geolib": "^2.0.24",
     "immutable": "^3.8.2",
+    "is-react": "^1.1.1",
```
@betodealmeida @mistercrunch was it necessary to add this dependency for functionality that is already provided by React natively? 

from the [source](https://github.com/treyhuffine/is-react/blob/master/index.js#L32): 
`React.isValidElement(typeElement)`

williaster replied:
FYI this added react 16 as a direct dependency (not a peer dependency), which means we have 2 versions of react floating around. it'd be great to get rid of this.

betodealmeida replied:
Let me take a look.

betodealmeida replied:
Fixed in https://github.com/apache/incubator-superset/pull/5478.

betodealmeida commented on superset/viz.py:
```diff
@@ -1219,24 +1205,50 @@ def run_extra_queries(self):
+                    self.to_series(
+                        diff, classed='time-shift-{}'.format(i), title_suffix=label))
+
+        return sorted(chart_data, key=lambda x: tuple(x['key']))
```
Taking a look at this.

betodealmeida replied:
IIRC this is needed in order to color the groups in different timeshifts with the same color (eg, make sure "boys" is red in the main series as well as all offsets).
